### ISSUE TITLE
"Placeholder too short" in hdf5-1.8.17-vc9_2

### ISSUE THREAD

**Issue by mewahl** [conda-forge/geopandas-feedstock#11]
Hello folks,

I ran into an error while trying to install geopandas in Windows. My environment's name ("geo") resulted in a destination name ("C:/Anaconda3/envs/_geo") that was too long for the placeholder present in hdf5 ("C:/conda/envs/_build"):

```
[geo] conda install -y -c conda-forge geopandas
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ......
Solving package specifications: ..............................................................
Package plan for installation in environment C:\Anaconda3\envs\geo:

The following NEW packages will be INSTALLED:

    fiona:       1.6.4-np111py27_1
    freetype:    2.6.3-vc9_1             [vc9]
    functools32: 3.2.3.2-py27_1
    gdal:        1.11.4-np111py27_vc9_10 [vc9]
    geopandas:   0.2-py27_0
    hdf5:        1.8.17-vc9_2            [vc9]
    libnetcdf:   4.4.0-vc9_1             [vc9]
    libpng:      1.6.23-vc9_0            [vc9]
    libtiff:     4.0.6-vc9_6             [vc9]
    matplotlib:  1.5.2-np111py27_4
    openjpeg:    2.1.0-vc9_5             [vc9]
    pandas:      0.18.1-np111py27_0
    proj.4:      4.9.2-py27_vc9_0        [vc9]
    pyqt:        4.10.4-py27_1
    rtree:       0.8.2-py27_0
    scipy:       0.17.1-np111py27_1
    shapely:     1.5.16-np111py27_0

Linking packages ...
Error: ERROR: placeholder 'C:/conda/envs/_build' too short in: hdf5-1.8.17-vc9_2
```

I was able to get around the problem by using a one-letter environment name, but I'm sure others will run into trouble. Is it any help to mention the issue here, given that the problem appears to lie with hdf5?

**Comment by ocefpaf** [comment]
I never saw that issue before,but I suspect it is not related to `hdf5` either. What version of `conda` and `conda-env` are you using? Can you send the result of `conda list` here so we can check that?

Ping @msarahan who might have an idea.

**Comment by mewahl** [comment]
Hi ocefpaf,

I hadn't seen this before either, but encountered similar "placeholder too short" issues for other packages ([like this one](http://stackoverflow.com/questions/33107872/placeholder-too-short-error-during-anaconda-installation-of-ncurses)) when I googled the error message. Apparently it is not rare for packages to use binary placeholders, but usually developers are careful to choose a longer string.

The error occurs in a fresh environment (conda version is 3.19.1; unsure how to check conda-env version). Full example below.

```
conda create -n nametoolong python=2.7
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ....
Solving package specifications: ........
Package plan for installation in environment C:\Anaconda3\envs\nametoolong:

The following NEW packages will be INSTALLED:

    pip:            8.1.2-py27_0
    python:         2.7.12-0
    setuptools:     23.0.0-py27_0
    vs2008_runtime: 9.00.30729.1-2
    wheel:          0.29.0-py27_0

Proceed ([y]/n)? y

Linking packages ...
[      COMPLETE      ]|##################################################| 100%
#
# To activate this environment, use:
# > activate nametoolong
#

activate nametoolong
Deactivating environment "C:\Anaconda3"...
Activating environment "C:\Anaconda3\envs\nametoolong"...

[nametoolong] conda -V
conda 3.19.1

[nametoolong] conda list
# packages in environment at C:\Anaconda3\envs\nametoolong:
#
pip                       8.1.2                    py27_0
python                    2.7.12                        0
setuptools                23.0.0                   py27_0
vs2008_runtime            9.00.30729.1                  2
wheel                     0.29.0                   py27_0

[nametoolong] conda install -y -c conda-forge geopandas
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ......
Solving package specifications: ...............................................................
Package plan for installation in environment C:\Anaconda3\envs\nametoolong:

The following NEW packages will be INSTALLED:

    click:           6.6-py27_0
    click-plugins:   1.0.3-py27_0
    cligj:           0.4.0-py27_0
    curl:            7.49.0-vc9_0            [vc9]
    cycler:          0.10.0-py27_0
    descartes:       1.0.2-py27_0
    expat:           2.1.0-vc9_1             [vc9]
    fiona:           1.6.4-np111py27_1
    freetype:        2.6.3-vc9_1             [vc9]
    functools32:     3.2.3.2-py27_1
    gdal:            1.11.4-np111py27_vc9_10 [vc9]
    geopandas:       0.2-py27_0
    geopy:           1.10.0-py27_0
    geos:            3.4.2-vc9_2             [vc9]
    hdf4:            4.2.11-vc9_4            [vc9]
    hdf5:            1.8.17-vc9_2            [vc9]
    jpeg:            9b-vc9_0                [vc9]
    krb5:            1.14.2-vc9_0            [vc9]
    libnetcdf:       4.4.0-vc9_1             [vc9]
    libpng:          1.6.23-vc9_0            [vc9]
    libspatialindex: 1.8.5-vc9_1             [vc9]
    libtiff:         4.0.6-vc9_6             [vc9]
    matplotlib:      1.5.2-np111py27_4
    mkl:             11.3.3-1
    munch:           2.0.4-py27_0
    numpy:           1.11.1-py27_0
    openjpeg:        2.1.0-vc9_5             [vc9]
    pandas:          0.18.1-np111py27_0
    proj.4:          4.9.2-py27_vc9_0        [vc9]
    pyparsing:       2.1.4-py27_0
    pyproj:          1.9.5.1-py27_0
    pyqt:            4.10.4-py27_1
    pysal:           1.11.1-py27_0
    python-dateutil: 2.5.3-py27_0
    pytz:            2016.4-py27_0
    rtree:           0.8.2-py27_0
    scipy:           0.17.1-np111py27_1
    shapely:         1.5.16-np111py27_0
    six:             1.10.0-py27_0
    sqlalchemy:      1.0.13-py27_0
    xerces-c:        3.1.2-vc9_0             [vc9]
    zlib:            1.2.8-vc9_3             [vc9]

Linking packages ...
Error: ERROR: placeholder 'C:/conda/envs/_build' too short in: hdf5-1.8.17-vc9_2
```

**Comment by ocefpaf** [comment]
Like  I suspect the problem is how conda encodes the place holder. @msarahan do we need to re-build `hdf5` with latest conda? Or that can be solved locally just by updating it before creating the env?

**Comment by msarahan** [comment]
You need to rebuild hdf5 with one of the two options here: http://conda.pydata.org/docs/building/meta-yaml.html#detect-binary-files-with-prefix - this will get you up to 80 characters.

Prefixes of 255 characters (bytes) will be supported in conda-build 2.0.  I have not had time yet to implement @pelson's suggestion to make this code fail gracefully on encountering a short prefix rather than crash.  Keep in min

**Comment by ocefpaf** [comment]
I guess are already at the limit then b/c `detect_binary_files_with_prefix` is already used. See 
https://github.com/conda-forge/hdf5-feedstock/blob/master/recipe/meta.yaml#L28

@mewahl sorry. I guess we will need to wait for `conda-build 2.0`.

**Comment by msarahan** [comment]
No, it isn't working right somehow.

> Error: ERROR: placeholder 'C:/conda/envs/_build' too short in: hdf5-1.8.17-vc9_2

that should be something like _build_placehold_placehold...

**Comment by ocefpaf** [comment]
I will try the second option and see how it goes then.

**Comment by rsignell-usgs** [comment]
@mewahl, I was having this problem on Windows also, but I just upgraded to the latest conda and I don't seem to have it anymore.  Can you try this and report back if it fixes the problem for you also?

```
(C:\)  conda update conda
(C:\)  conda -V
conda 4.1.11
```

**Comment by dharhas** [comment]
So this is still broken on OS X

if I have conda install at /Users/dharhas/misc/builds/cmb/build-sb-capstone/install/conda/

and then try:

conda create -c conda-forge -n very-very-long-name python=3 hdf5

it fails on ncurses

Error: ERROR: placeholder '/Users/travis/miniconda3/envs/_build_placehold_placehold_placehold_placehold_pla' too short in: conda-forge::ncurses-5.9-8

**Comment by ocefpaf** [comment]
What version of conda do you have there @dharhas? We are not sure why the problem went away on Windows with `conda 4.1.11`, but it did...

**Comment by dharhas** [comment]
conda 4.1.11

**Comment by ocefpaf** [comment]
Alright. Not sure what happened on Window then, but they are in luck. The rest of us need to wait for `conda-build 2.0` (ping @msarahan :wink:)

**Comment by darribas** [comment]
Hello,

I think I'm having a similar issue with this on Appveyor. You can have a look at the log here:

https://ci.appveyor.com/project/darribas/gds16/build/job/oy3vk0me77sfxcdu

It's weird because I'm pretty sure it's using the latest conda and a `conda-build>2`.

Any ideas or suggestions would be most appreciated. Thanks!

**Comment by darribas** [comment]
Super! Thanks for the answer @ocefpaf . I'm teaching a course that starts next week to 100 students, do you think it'll be working by then? Thank you so much for the incredible service conda-forge does, it makes things like my course possible!!!

**Comment by ocefpaf** [comment]
>  I'm teaching a course that starts next week to 100 students, do you think it'll be working by then?

probably not... Sorry. the solution demands a major re-build of many package. You will have to test your env to create a mix of the default channel `HDF5` and `conda-forge`'s geopandas. I will see if I can put one together.

**Comment by darribas** [comment]
Thanks very much! An example would be most useful!!!

**Comment by ocefpaf** [comment]
I would start with something like this

```yaml
name: geopandas
channels:
    - defaults
    - conda-forge
dependencies:
    - python=2.7
    - geopandas
```

and then keep adding the dependencies and checking if you need to pin any version. Note that I placed `defaults` on top of `conda-forge` there to use the channel preference feature of the recent `conda` releases. So make sure your `conda` is updated!

Ultimately you do not want to get `gdal` and `hdf5` from conda-forge. At least until we rebuild everything.

PS: `defaults` is also doing a re-build. Hopefully they are ahead of `conda-forge` and you won't have trouble with their packages.

**Comment by darribas** [comment]
Same problem :(

https://ci.appveyor.com/project/darribas/gds16/build/job/tmngyh070s0vcpmw

Seems to still be picking up the `conda-forge` version instead of the `defaults`...

**Comment by darribas** [comment]
For convenience, here's the env yml file:

https://github.com/darribas/gds16/blob/master/content/infrastructure/install_gds_stack.yml

**Comment by ocefpaf** [comment]
First of all. Really nice repo!

Now that I have all the dependencies I will take a look if I can put a working env together. Not sure though b/c the GIS stack on conda-forge is highly dependent on many feedstocks that I lost control over to people who does not really care about GIS. Like `hdf5`, `numpy`, and `openblas`.

**Comment by ocefpaf** [comment]
@darribas I need to make more tests but this may work:

```yaml
name: gds
channels:
    - defaults
    - conda-forge
dependencies:
    - python=2.7
    - pip
    - bokeh
    - cartopy
    - clusterpy
    - dask
    - ipython
    - ipywidgets
    - jupyter
    - mplleaflet
    - networkx
    - palettable
    - qgrid
    - rasterio
    - scikit-learn
    - seaborn
    - statsmodels
    - xlrd
    - xlsxwriter
    # FIXME: Workaround packages incompatibilities.
    # remove once conda-forge rebuilds everything with conda-build 2.
    - hdf5 ==1.8.15.1
    - gdal
    - pandas
    - pyproj
    - shapely
    - fiona
    - six
    - rtree
    - descartes
    - psycopg2
    - sqlalchemy
    - geopy ==1.10.0
    - pysal
    - pip:
        - geopandas
```

I can put a PR together later to test and tweak it for you but I need to go now and I will be back only in the afternoon.

**Comment by ocefpaf** [comment]
That creates a broken env b/c of the outdated software in `defaults` :unamused: 

You are between a rock and a hard place here...

The GIS stack needs more love from both `conda-forge` and `defaults`. Ideally continuum should pull the modern packages (`gdal`, `fiona`, `rasterio`, etc) into `defaults` so `conda-forge` can be the "latest" unstable and `defaults` the reliable "stable" channel. However, right now, both are neither :unamused:

**Comment by jorisvandenbossche** [comment]
> so conda-forge can be the "latest" unstable and defaults the reliable "stable" channel. However, right now, both are neither

Is there a reason that conda forge cannot be this reliable stable channel for now? Or at least have the versions needed for such a stable set-up. (but the answer is maybe just: time ..)

**Comment by ocefpaf** [comment]
> Is there a reason that conda forge cannot be this reliable stable channel for now? Or at least have the versions needed for such a stable set-up. (but the answer is maybe just: time ..)

That is a good question! I try my best to keep my feedstocks stable but I cannot prevent people, with more free time and higher "noise level", to just go ahead and make changes that breaks the stack. I gave up fighting that after the `gdal` fiasco when someone forced a single-threaded openblas without any regard the the Met/Ocean/GIS community. That change broke everything that depended on numpy+gdal and I had to deal with 3-months of broken packages and complaints until we fixed `gdal` to work with this person many breaking changes.

We introduced the enhancement proposals to reduce the change of that happening again but unless the [community](https://conda-forge.hackpad.com/conda-forge-meetings-2YkV96cvxPG) starts participating more in conda-forge there is very little I can do alone.
